hep/proposals/0000-decorators.md

```diff
@@ -0,0 +1,299 @@
+# Meta
+[meta]: #meta
+- Name: Hera Decorator Syntax
+- Start Date: 2024-04-13
+- Author(s): [@samj1912](https://github.com/elliotgunton), [@elliotgunton](https://github.com/elliotgunton)
+- Supersedes: (put "N/A" unless this replaces an existing HEP, then link to that HEP)
+
+# Table of Contents
+[table-of-contents]: #table-of-contents
+- [Meta](#meta)
+- [Table of Contents](#table-of-contents)
+- [Overview](#overview)
+- [Motivation](#motivation)
+- [Proposal](#proposal)
+- [Code Examples](#code-examples)
+  - [Basic Workflow](#basic-workflow)
+  - [DAG Template](#dag-template)
+  - [Steps Template](#steps-template)
+  - [Container Template](#container-template)
+  - [Workflow Template Ref](#workflow-template-ref)
+  - [How to teach (OPTIONAL)](#how-to-teach-optional)
+- [Implementation (OPTIONAL)](#implementation-optional)
+  - [Link to the Implementation PR](#link-to-the-implementation-pr)
+- [Migration (OPTIONAL)](#migration-optional)
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+- [Prior Art](#prior-art)
+- [Unresolved Questions (OPTIONAL)](#unresolved-questions-optional)
+
+# Overview
+[overview]: #overview
+
+This HEP introduces a new and powerful decorator based syntax for defining all kinds of Workflows and Templates.
+
+# Motivation
+[motivation]: #motivation
+
+The current context manager based syntax is good but it doesn't map well to template types that contain inputs/outputs. With the `@script` decorator, we have made hera feel extremely Pythonic and it has the added benefit of being runnable locally. We wish to extend this syntax to other template types which will allow for better readability and local testing and it will also standardize the template interface for Argo Workflows.
+
+# Proposal
+
+This HEP proposes that every template be mapped into a decorated function with HeraIO classes as the inputs and outputs.
+
+# Code Examples
+
+## Basic Workflow
+
+```python
+from hera.workflows import WorkflowTemplate
+from hera import io as hio
+
+# We start by defining our Workflow Template
+wt = WorkflowTemplate(name="my-template")
+
+# This defines the a template input
+class MyInput(hio.Input):
+    user: str
+
+@wt.entrypoint  # Sets hello_world as the default entrypoint for the workflow template
+@wt.script  # Adds a new script template to the workflow template called hello_world
+def hello_world(my_input: MyInput) -> hio.Output:
```

Comment:
I would think these `wt.scripts` _have_ to use the runner? Omitted for brevity or should `constructor="runner"` be somewhere in the code (and called out in the HEP)?

Reply:
The decorator via the template class needs to enforce it. There are no other options allowed.